middle_end/flambda2/simplify/simplify_switch_expr.ml

```diff
@@ -210,15 +205,169 @@ let find_cse_simple dacc required_names prim =
+      (Compilation_unit.get_current_exn ())
+      (Linkage_name.of_string (Variable.unique_name var))
+  in
+  let uacc =
+    let fields = List.map Field_of_static_block.tagged_immediate consts in
+    UA.add_lifted_constant uacc
+      (LC.create_definition
+         (LC.Definition.block_like
+            (DA.denv dacc_before_switch)
+            block_sym T.any_block ~symbol_projections:Variable.Map.empty
+            (RSC.create_block rebuilding tag Immutable ~fields)))
+  in
+  (* CR mshinwell: consider sharing the constants *)
+  let block = Simple.symbol block_sym in
+  let access_kind : P.Block_access_kind.t =
+    Values
+      { tag = Known tag;
+        size = Known (TI.of_int num_consts);
+        field_kind = Immediate
+      }
+  in
+  let load_from_block =
+    Named.create_prim
+      (Binary (Block_load (access_kind, Immutable), block, tagged_scrutinee))
+      dbg
+  in
+  let arg_var = Variable.create "arg" in
+  let arg = Simple.var arg_var in
+  let final_arg_var, final_arg =
+    match must_untag_lookup_table_result with
+    | Must_untag ->
+      let final_arg_var = Variable.create "final_arg" in
+      final_arg_var, Simple.var final_arg_var
+    | Leave_as_tagged_immediate -> arg_var, arg
+  in
+  (* Note that, unlike for the untagging of normal Switch scrutinees, there's no
+     problem with CSE and Data_flow here. The reason is that in this case the
+     generated primitive always names a fresh variable, so it will never be
+     eligible for CSE. *)
+  (* CR mshinwell: we could probably expose the actual integer counts of
+     continuations in [Name_occurrences] and then try to inline out [dest]. This
+     might happen anyway in the backend though so this probably isn't that
+     important for now. *)
+  let apply_cont = Apply_cont.create dest ~args:[final_arg] ~dbg in
+  let free_names_of_body = Apply_cont.free_names apply_cont in
+  let untag_arg_prim : P.t = Unary (Untag_immediate, arg) in
+  let expr =
+    let body =
+      let body = RE.create_apply_cont apply_cont in
+      match must_untag_lookup_table_result with
+      | Leave_as_tagged_immediate -> body
+      | Must_untag ->
+        let bound = BPt.singleton (BV.create final_arg_var NM.normal) in
+        let untag_arg = Named.create_prim untag_arg_prim dbg in
+        RE.create_let rebuilding bound untag_arg ~body ~free_names_of_body
+    in
+    let bound = BPt.singleton (BV.create arg_var NM.normal) in
+    RE.create_let rebuilding bound load_from_block ~body ~free_names_of_body
+  in
+  let extra_free_names =
+    NO.union
+      (Named.free_names load_from_block)
+      (NO.remove_var free_names_of_body ~var:final_arg_var)
+  in
+  let increase_in_code_size =
+    (* Very likely negative. *)
+    Code_size.( - )
+      (Code_size.( + )
+         (Code_size.apply_cont apply_cont)
+         (match must_untag_lookup_table_result with
+         | Must_untag -> Code_size.prim untag_arg_prim
+         | Leave_as_tagged_immediate -> Code_size.zero))
```

Comment:
Shouldn't you also add the cost of the block load  ?

Reply:
Fixed